### PR TITLE
linux-firmware: remove fod usage

### PR DESCRIPTION
**Pull request by SuperSandro2000**
It was re-introduced in https://github.com/NixOS/nixpkgs/pull/59748 without an explanation and the original PR https://github.com/NixOS/nixpkgs/pull/44605 is not available anymore.

Removing FOD makes using overlays easier and avoids potential issues when something in the inputs changes which causes a FOD change.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by fpletz** [comment]
> Removing FOD makes using overlays easier and avoids potential issues when something in the inputs changes which causes a FOD change.

Well, the idea is that even if the hash changes because the inputs change, the FOD doesn't have to be rebuilt because the actual output doesn't change (what @K900 said). The idea back then was to save resources on rebuilds, space and traffic both on NixOS infrastructure and on NixOS installations.

> Removing FOD makes using overlays easier and avoids potential issues when something in the inputs changes which causes a FOD change.

With overlays you mean people overriding the derivation also needing to update the hash? That's a valid UX concern.

I don't have an opinion either way since overriding linux-firmware shouldn't be necessary as one can simply add new firmware packages to `hardware.firmware`.

**Comment by SuperSandro2000** [comment]
> With overlays you mean people overriding the derivation also needing to update the hash? That's a valid UX concern.

We observed that today on the matrix channel and I took this as an opportunity to remove it.

> I don't have an opinion either way since overriding linux-firmware shouldn't be necessary as one can simply add new firmware packages to `hardware.firmware`.

I could see recommending that to someone on stable who needs newer firmware for a new laptop and would be very confused why it would not work.

**Comment by K900** [comment]
We should really be backporting linux-firmware to stable, now that I think about it.

**Review by fpletz — COMMENTED** [review]
> I could see recommending that to someone on stable who needs newer firmware for a new laptop and would be very confused why it would not work.

Since overriding requires some tinkering and reading the actual code anyways, I don't see how people could not see the FOD except if they are new to Nix and then they will learn something. After pondering about it some more, I don't think the UX issue justifies this change. This is what FODs are for after all.

If somebody ends up merging this PR, remove me from maintainers and add yourself instead.

File: pkgs/os-specific/linux/firmware/linux-firmware/default.nix [hunk 1]
```diff
@@ -1,29 +1,23 @@
-let
-  source = import ./source.nix;
```
Comment (ajs124) [inline pkgs/os-specific/linux/firmware/linux-firmware/default.nix:2]: shouldn't this also delete the `source.nix`, since it's now unused

**Comment by SuperSandro2000** [comment]
> We should really be backporting linux-firmware to stable, now that I think about it.

Lets leave that discussion open for now since it is a bit unrelated to this PR.

---

I also tested the updater and to my testing it works.

**Comment by SuperSandro2000** [comment]
We should just merge this, since it is a micro optimization which has little to no impact if rebuild but makes doing overlays unnecessarily harder.

**Pull request closed without merging**